### PR TITLE
Fix error message for Publish-PSResource for MyGet.org feeds

### LINKED ISSUE #1246: Ensure Publish works for all V3 repositories
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PSResourceGet)
- [X] Search the existing issues.

### Steps to reproduce

Publish-PSResource works for:

GithubPackages (it'll publish as a private package to the namespace specified, and then one can link it to the desired repository and change visibility)
JFrog Artifactory

Publish does not work for:

MyGet feed gives 403 (Forbidden) error
NuGet.org gives 401 error (but it's about missing the same key as MyGet does from the header)
AzureArtifacts gives 400 (Bad Request) error

### Expected behavior

```console
Publish-PSResource should work for all V3 repositories we support.
```

### Actual behavior

```console
Azure Devops, NuGet.org, and MyGet do not publish without errors.
```

### Error details

_No response_

### Environment data

```powershell
latest PSResource Get version, latest PSVersion
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by anamnavi**
# PR Summary

Improve error messages presented when running Publish-PSResource for MyGet.org repository. When publishing to this type of repository errors can be written out when the -ApiKey value is missing or incorrect. The server side error message is not clear enough on its own so add to it.

## PR Context

Resolves #1246 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/PublishPSResource.cs [hunk 1]
```diff
@@ -917,24 +917,35 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
                     if (e.Message.Contains("API"))
                     {
                         // For PSGallery when ApiKey is not provided.
-                        var message = String.Format("Repository '{0}': {1} Please try running again with the -ApiKey parameter and specific API key for the repository specified.", repoName, e.Message);
+                        var message = String.Format("Repository '{0}': Please try running again with the -ApiKey parameter and specific API key for the repository specified. Exception: {1}", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message = String.Format("Repository '{0}': Please try running again with the -ApiKey parameter and specific API key for the repository specified. Exception: '{1}'", repoName, e.Message);

```

File: src/code/PublishPSResource.cs [hunk 2]
```diff
@@ -917,24 +917,35 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
                         ex = new ArgumentException(message);
                         var ApiKeyError = new ErrorRecord(ex, "401ApiKeyError", ErrorCategory.AuthenticationError, null);
                         error = ApiKeyError;
                     }
                     else
                     {
                         // For ADO repository feeds that are public feeds, when the credentials are incorrect.
-                        var message = String.Format("Repository '{0}': {1} The Credential provided was incorrect.", repoName, e.Message);
+                        var message = String.Format("Repository '{0}': The Credential provided was incorrect. Exception: {1}", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message = String.Format("Repository '{0}': The Credential provided was incorrect. Exception: '{1}'", repoName, e.Message);

```

File: src/code/PublishPSResource.cs [hunk 3]
```diff
@@ -917,24 +917,35 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
                     }
                 }
                 else if (e.Message.Contains("403"))
                 {
-                    var Error403 = new ErrorRecord(ex, "403Error", ErrorCategory.PermissionDenied, null);
-                    error = Error403;
+                    if (repoUri.Contains("myget.org"))
+                    {
+                        // For myGet.org repository feeds when the ApiKey is missing or incorrect.
+                        var message = String.Format("Repository '{0}': The ApiKey provided is incorrect or missing. Please try running again with the -ApiKey parameter and correct API key value for the repository. Exception: {1}", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message = String.Format("Repository '{0}': The ApiKey provided is incorrect or missing. Please try running again with the -ApiKey parameter and correct API key value for the repository. Exception: '{1}'", repoName, e.Message);

```

File: src/code/PublishPSResource.cs [hunk 4]
```diff
@@ -954,7 +965,7 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
                 //  for ADO repository feeds that are private feeds the error thrown is different and the 401 is in the inner exception message
                 if (e.InnerException.Message.Contains("401"))
                 {
-                    var message = String.Format ("Repository '{0}': {1} The Credential provided was incorrect.", repoName, e.InnerException.Message);
+                    var message = String.Format ("Repository '{0}': The Credential provided was incorrect. Exception {1}", repoName, e.InnerException.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                    var message = String.Format ("Repository '{0}': The Credential provided was incorrect. Exception '{1};", repoName, e.InnerException.Message);

```

File: src/code/PublishPSResource.cs [hunk 5]
```diff
@@ -917,24 +917,35 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
                         error = Error401;
                     }
                 }
                 else if (e.Message.Contains("403"))
                 {
-                    var Error403 = new ErrorRecord(ex, "403Error", ErrorCategory.PermissionDenied, null);
-                    error = Error403;
+                    if (repoUri.Contains("myget.org"))
+                    {
+                        // For myGet.org repository feeds when the ApiKey is missing or incorrect.
```
Comment (alerickson) [inline src/code/PublishPSResource.cs:942]: Is MyGet the only one that has this error when the ApiKey is missing/incorrect?

**Review by alerickson — CHANGES_REQUESTED** [review]
Just a question and some quotation suggestions

File: src/code/PublishPSResource.cs (hunk 5 above)
Comment (anamnavi) [reply src/code/PublishPSResource.cs:942]: Yes, only MyGet throws a 403 when ApiKey is missing, the other repositories we have documented throw a 400 or 401.

Comment (anamnavi) [reply src/code/PublishPSResource.cs:942]: JFrog will throw 403 if ApiKey is specified (correct or incorrect)

File: src/code/PublishPSResource.cs [hunk 6]
```diff
@@ -917,24 +917,47 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
                     if (e.Message.Contains("API"))
                     {
                         // For PSGallery when ApiKey is not provided.
-                        var message = String.Format("Repository '{0}': {1} Please try running again with the -ApiKey parameter and specific API key for the repository specified.", repoName, e.Message);
+                        var message = String.Format("Repository '{0}': Please try running again with the -ApiKey parameter and specific API key for the repository specified. Exception: '{1}'", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message = $"Could not publish to repository '{repoName}'. Please try running again with the -ApiKey parameter and the API key for the repository specified. Exception: '{e.Message}'";

```

File: src/code/PublishPSResource.cs [hunk 7]
```diff
@@ -917,24 +917,47 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
+
                         ex = new ArgumentException(message);
                         var ApiKeyError = new ErrorRecord(ex, "401ApiKeyError", ErrorCategory.AuthenticationError, null);
                         error = ApiKeyError;
                     }
                     else
                     {
                         // For ADO repository feeds that are public feeds, when the credentials are incorrect.
-                        var message = String.Format("Repository '{0}': {1} The Credential provided was incorrect.", repoName, e.Message);
+                        var message = String.Format("Repository '{0}': The Credential provided was incorrect. Exception: '{1}'", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message =$"Could not publish to repository '{repoName}'. The Credential provided was incorrect. Exception: '{e.Message}'";

```

File: src/code/PublishPSResource.cs [hunk 8]
```diff
@@ -917,24 +917,47 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
                     }
                 }
                 else if (e.Message.Contains("403"))
                 {
-                    var Error403 = new ErrorRecord(ex, "403Error", ErrorCategory.PermissionDenied, null);
-                    error = Error403;
+                    if (repoUri.Contains("myget.org"))
+                    {
+                        // For myGet.org repository feeds when the ApiKey is missing or incorrect.
+                        var message = String.Format("Repository '{0}': The ApiKey provided is incorrect or missing. Please try running again with the -ApiKey parameter and correct API key value for the repository. Exception: '{1}'", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message = $"Could not publish to repository '{repoName}'. The ApiKey provided is incorrect or missing. Please try running again with the -ApiKey parameter and correct API key value for the repository. Exception: '{e.Message}'";

```

File: src/code/PublishPSResource.cs [hunk 9]
```diff
@@ -954,7 +977,8 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
                 //  for ADO repository feeds that are private feeds the error thrown is different and the 401 is in the inner exception message
                 if (e.InnerException.Message.Contains("401"))
                 {
-                    var message = String.Format ("Repository '{0}': {1} The Credential provided was incorrect.", repoName, e.InnerException.Message);
+                    var message = String.Format ("Repository '{0}': The Credential provided was incorrect. Exception '{1};", repoName, e.InnerException.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                    var message = $"Could not publish to repository '{repoName}'. The Credential provided was incorrect. Exception '{e.InnerException.Message}'";

```

File: src/code/PublishPSResource.cs [hunk 10]
```diff
@@ -917,24 +917,47 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
+                        var message = String.Format("Repository '{0}': The ApiKey provided is incorrect or missing. Please try running again with the -ApiKey parameter and correct API key value for the repository. Exception: '{1}'", repoName, e.Message);
+
+                        ex = new ArgumentException(message);
+                        var Error403 = new ErrorRecord(ex, "403Error", ErrorCategory.PermissionDenied, null);
+                        error = Error403;
+                    }
+                    else if (repoUri.Contains(".jfrog.io"))
+                    {
+                        // For JFrog Artifactory repository feeds when the ApiKey is provided, whether correct or incorrect, as JFrog does not require -ApiKey (but does require ApiKey to be present as password to -Credential).
+                        var message = String.Format("Repository '{0}': The ApiKey provided is not needed for JFrog Artifactory. Please try running again without the -ApiKey parameter but ensure that -Credential is provided with ApiKey as password. Exception: '{1}'", repoName, e.Message);
```
Comment (alerickson) [inline src/code/PublishPSResource.cs]:
```suggestion
                        var message = $"Could not publish to repository '{repoName}'. The ApiKey provided is not needed for JFrog Artifactory. Please try running again without the -ApiKey parameter but ensure that -Credential is provided with ApiKey as password. Exception: '{e.Message}'";

```

**Review by alerickson — APPROVED** [review]
Just some formatting and wording suggestions.

**Pull request merged**